### DISCUSSION TITLE
Proper way to register custom type in store state?

### DISCUSSION THREAD

**Discussion by jdheeter** [vuejs/pinia#1178] [Help and Questions] [answered]
Most of the Pinia examples are using simple primitive types. When using more complex custom types I'm finding that the type returned by the pinia store instance is "exploded" and no longer considered to be equal to the type assigned to the state which is defined in the store.

![image](https://user-images.githubusercontent.com/2530957/160976370-1cb55e3d-64d5-4d18-8a56-d6a85c9a1e75.png)
![image](https://user-images.githubusercontent.com/2530957/160976389-bf5a01fa-8c35-49f8-883d-681e490883b6.png)
So for this pinia store, I would expect that `tables.packs.cartoon` is of type `Packs[]`

However when used in a .vue file
![image](https://user-images.githubusercontent.com/2530957/160976514-0a7e927e-9f0d-44b7-8605-9dbe90dbfeab.png)
![image](https://user-images.githubusercontent.com/2530957/160976534-35eaad89-fc30-4ac6-bcf2-c06555b65ca7.png)
![image](https://user-images.githubusercontent.com/2530957/160976550-7cc69903-dca2-46ad-a8a9-01b92d84f94f.png)

So all that gibberish is the actual type definition of `Packs[]`
and now `tables.packs.cartoon` is no longer of type `Packs[]`

<img width="824" alt="image" src="https://user-images.githubusercontent.com/2530957/160976779-73e2e761-d8ef-49d6-a447-94f1e9d12cd2.png">
Why is the type exploded open like this? It should simply be of type `Packs[]`

Is Pinia meant to only be used with basic primitives?

Here is the definition of the `Packs` type and `Struct`
```js
@Struct.type('packs')
export class Packs extends Struct {
    @Struct.field(UInt64) template_id!: UInt64
    @Struct.field(Asset) base_price!: Asset
    @Struct.field(Asset) floor_price!: Asset
    @Struct.field(TimePointSec) last_sold!: TimePointSec
    @Struct.field('string') pack_name!: string
}
```
```js
declare class Struct implements ABISerializableObject {
    static abiName: string;
    static abiFields: ABIField[];
    static abiBase: ABISerializableConstructor;
    static from<T extends StructConstructor>(this: T, value: any): InstanceType<T>;
    static from(value: any): unknown;
    static get structFields(): ABIField[];
    /** @internal */
    constructor(object: any);
    /**
     * Return true if this struct equals the other.
     *
     * Note: This compares the ABI encoded bytes of both structs, subclasses
     *       should implement their own fast equality check when possible.
     */
    equals(other: any): boolean;
    /** @internal */
    toJSON(): any;
}
declare namespace Struct {
    function type(name: string): <T extends StructConstructor>(struct: T) => T;
    function field(type: ABISerializableConstructor | string, options?: Partial<ABIField>): <T extends Struct>(target: T, name: string) => void;
}
```

**Comment by LinusBorg** [comment, marked as answer]
As a general advice:

Vue's reactivity work best with primitive data, that's true for both runtime and the types. And generally, it's also recommended to keep complex objects with side effects or complex internal state non-reactive, as you might get unintented side-effects when you make these object's properties reactive, depending on the internal interactions.

This specific issue is, as far as I can tell from the information provided, because of Generics not working too well with Vue's (type) unwrapping of refs when the generic could be a ref.

I can't pin it down for your types here as I don't know them all (i.e.: what's `Int`?), but generally, if a generic's specific value could potentially contain a ref(), then there can be a conflict because:

```js
const x = { foo: ref('bar') }
// => shape: { foo: Ref<string> }

const reactiveX = reactive(x)
// => shape: { foo: string }
```

There are generally two ways to solve this depending on the requirements: 

1. Typecasting
2. `markRaw()` to keep Vue from making complex objects reactive in the first place.
3. (2b) the whole state could maybe also be wrapped in a `shallowReactive()` which would make all root state properties reactive, but not their nested objects.

this is not a Pinia problem, ~~so should be closed here~~ (didn't realize this was a topic, not an issue ^^).

**Reply by jdheeter** [reply]
Very clear response. I wasn't aware of the utilities to limit reactivity on nested properties. Thanks!

**Reply by miquelromero** [reply]
Thanks for your clarifications. Could we expect in a future a way of getting the original types, or we should assume that we will have to typecast non-primitive types in pinia forever?
